tests/python/multidevice/test_communication.py

```diff
@@ -5,133 +5,156 @@
-    d = multidevice_test.size
-    mesh = nvfuser.DeviceMesh(range(d))
-
-    class Model(FusionDefinition):
-        def definition(self):
-            self.inp = self.define_tensor(
-                (-1, -1, -1), contiguity=True, dtype=DataType.Float
-            )
-            out = self.ops.sum(self.inp, [1])
-            self.add_output(out)
-
-        def multidevice_schedule(self):
-            self.sched._set_device_mesh(self.inp, mesh)
-            self.sched.split(self.inp, 1, d, False)
-            self.sched.parallelize(self.inp, 1, nvfuser.ParallelType.mesh_x)
+def test_allreduce(multidevice_direct_test):
+    d = multidevice_direct_test.size
+    mesh = nvfuser.multidevice.DeviceMesh(range(d))
+
+    def _definition(fd: FusionDefinition):
+        inp = fd.define_tensor((-1, -1, -1), contiguity=True, dtype=DataType.Float)
+        out = fd.ops.sum(inp, [1])
+        fd.add_output(out)
+
+    def _multidevice_schedule(fd: FusionDefinition):
+        for inp in fd.fusion.inputs():
+            inp.set_device_mesh(mesh)
+            inp.split(1, d, inner_split=False)
+            inp.axis(1).parallelize(nvfuser.ParallelType.mesh_x)
 
     m = 2
     k = d * 3
     n = 5
     unsharded = torch.randn(m, k, n)
-    sharded = multidevice_test.shard_tensor(unsharded, 1, mesh)
-
-    fd = Model()
-    outputs, _ = fd.execute([sharded])
-    torch.testing.assert_close(outputs[0].cpu(), unsharded.sum(1))
-
+    sharded = multidevice_direct_test.shard_tensor(unsharded, 1, mesh)
 
-@pytest.mark.mpi
-def test_reduce_scatter(multidevice_test):
-    d = multidevice_test.size
-    mesh = nvfuser.DeviceMesh(range(d))
-
-    class Model(FusionDefinition):
-        def definition(self):
-            self.inp = self.define_tensor(
-                (d, d * 4), contiguity=True, dtype=DataType.Float
-            )
-            self.out = self.ops.sum(self.inp, [0])
-            self.add_output(self.out)
+    with FusionDefinition() as fd:
+        _definition(fd)
+        _multidevice_schedule(fd)
 
-        def multidevice_schedule(self):
-            self.sched._set_device_mesh(self.inp, mesh)
-            self.sched._set_device_mesh(self.out, mesh)
+    (output,) = fd.execute([sharded])
+    torch.testing.assert_close(output.cpu(), unsharded.sum(1))
 
-            self.sched.parallelize(self.inp, 0, nvfuser.ParallelType.mesh_x)
 
-            self.sched.split(self.out, -1, d, False)
-            self.sched.parallelize(self.out, -2, nvfuser.ParallelType.mesh_x)
+@pytest.mark.mpi
+def test_reduce_scatter(multidevice_direct_test):
+    d = multidevice_direct_test.size
+
+    # The first dimension of the TensorView is the world size.
+    # When it is one, the axis becomes a broadcast iterDomain.
+    # The sum operations uses squeeze for size-1 broadcast domains.
+    # Squeeze is not supported by getCommunicationInfo.
+    # This test is skipped when the world size is 1.
+    if d == 1:
+        pytest.skip("This test requires > 1 MPI processes")
```

Comment:
Was this case executing fine in the previous version? I don't see the test being skipped earlier based on devices.

Reply:
It threw an error too.